controllers/calendar/RecurringEventsController.php

```diff
@@ -0,0 +1,176 @@
+<?php
+
+namespace humhub\modules\smartVillage\controllers\calendar;
+
+use DateTime;
+use humhub\modules\calendar\interfaces\CalendarService;
+use humhub\modules\calendar\models\CalendarEntry;
+use humhub\modules\calendar\models\fullcalendar\FullCalendar;
+use humhub\modules\content\components\ActiveQueryContent;
+use humhub\modules\content\components\ContentActiveRecord;
+use humhub\modules\content\models\ContentContainer;
+use humhub\modules\rest\components\BaseContentController;
+use humhub\modules\rest\definitions\ContentDefinitions;
+
+class RecurringEventsController extends BaseContentController
+{
+    public static $moduleId = 'Calendar';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getContentActiveRecordClass()
+    {
+        return CalendarEntry::class;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function returnContentDefinition(ContentActiveRecord $contentRecord)
+    {
+        /** @var CalendarEntry $contentRecord */
+        return FullCalendar::getFullCalendarArray($contentRecord);
+    }
+
+    /**
+     * @return array
+     * @throws \Throwable
+     * Get all the recurring events lies between start_date and end_date
+     * We filter the events based on their title
+     */
+    public function actionIndex(){
+        $output = [];
+        if(!isset($_GET['start_date']) || !isset($_GET['end_date'])){
+            return $this->returnError(400,"start_date and end_date parameters are required!");
+        }
+
+        //Get all the calendar entries lies between the start_date and end_date
+        $calendarService = new CalendarService();
+        $entries = $calendarService->getCalendarItems(new DateTime($_GET['start_date']), new DateTime($_GET['end_date']));
+
+        //If the calendar entry not found in given date of range
+        if($entries == null){
+            return $this->returnError(404, "No recurring events are present between ".$_GET['start_date'] ." and ". $_GET['end_date']);
+        }
+        //Get those calendar events which are recurring
+        $calendarEntryTitle = [];
+        $calendarEntries = CalendarEntry::find()->all();
+        foreach($calendarEntries as $calendarEntry){
+            if(isset($calendarEntry->rrule)){
+                array_push($calendarEntryTitle,$calendarEntry['title']);
+            }
+        }
+
+        foreach ($entries as $entry) {
+            /** @var ContentActiveRecord $entry */
+            //return only recurring events
+            if(in_array($entry['title'],$calendarEntryTitle)){
+                $output[] = $this->returnContentDefinition($entry);
+            }
+        }
+        //If recurring event not found between given date range
+        if($output == null){
+            return $this->returnError(404, "No recurring events are present between ".$_GET['start_date'] ." and ". $_GET['end_date']);
+        }
+        return $output;
```

Comment:
when i compare the output of the default calendar entry endpoint and the recurring one, there is much difference in the json.

1. keys are different, e.g. `all_day` and `allDay` -> please make all keys the same as the original result from `/calendar`
2. a lot of keys and information are missing, for example `participants` and `participation_mode`, `description`, `content`
3. there are additional keys and values, what does they mean? do we need them? like `viewMode: "modal"` for example.

as we discussed in the issue i thought we will receive same results:

![image](https://user-images.githubusercontent.com/1942953/183886115-31ca99c7-32b1-4aca-9ea2-d0d3ced64a80.png)

how should we decide which keys to use from the api if we get calendar data? only `id` and `title` are the same and reusable.
example with dates:
original:
```
"start_datetime": "2022-06-23 10:00:00",
"end_datetime": "2022-08-23 00:00:00",
```

new:
```
"start": "2022-06-23T10:00:00+02:00",
"end": "2022-08-23T00:00:00+02:00",
```

not only the keys are different that need to be accessed but also the value format.

can you please compare the resulting data and adjust them accordingly based on the original calendar returns please.

Reply:
Hi @donni106 
We only add data of the recurring model in the body. you can check the original calendar API for creating events.

![image](https://user-images.githubusercontent.com/103625795/183889993-ea07ac7b-3a52-4daa-a9d8-175ce220a670.png)


![image](https://user-images.githubusercontent.com/103625795/183890145-36451b80-56b0-4203-8932-53838c7f946a.png)

The key name difference is by default

Reply:
i am not speaking about creating events but what the results are.

calendar entry:
```
{
  "id": 60,
  "title": "BbbB",
  "description": "",
  "start_datetime": "2022-06-23 10:00:00",
  "end_datetime": "2022-08-23 00:00:00",
  "all_day": 0,
  "participation_mode": 2,
  "color": "#107821",
  "allow_decline": 0,
  "allow_maybe": 0,
  "time_zone": "Europe/Berlin",
  "participant_info": "",
  "closed": 0,
  "max_participants": 0,
  "content": {
    "id": 266,
    "metadata": {
      "id": 266,
      "guid": "b4787956-12asdeb-4541-abeb-cbf4532eeee3",
      "object_model": "humhub\\modules\\calendar\\models\\CalendarEntry",
      "object_id": 60,
      "visibility": 1,
      "created_by": {
        "id": 11,
        "guid": "4853c569-e4asd1c-47c4-bfba-905611d18dbd",
        "display_name": "John Doe",
        "url": "http://..."
      },
      "created_at": "2022-08-10 12:30:28",
      "url": "/s/t..../calendar/entry/view-recurrence?id=60&parent_id=57&recurrence_id=20220623T100000",
      "contentcontainer_id": 52,
      "stream_channel": "default"
    },
    "comments": {
      "total": "0",
      "latest": []
    },
    "likes": {
      "total": 0
    },
    "topics": [],
    "files": []
  },
  "participants": {
    "attending": [],
    "maybe": [],
    "declined": []
  }
},
```

recurring calendar entry:
```
{
  "id": 49,
  "title": "Ww",
  "editable": false,
  "backgroundColor": "#107821",
  "allDay": false,
  "viewUrl": "/s/test-nur-auf-einladung/calendar/entry/view-recurrence?id=49&cal=1&parent_id=48&recurrence_id=20220623T100000",
  "viewMode": "modal",
  "icon": "fa-calendar",
  "start": "2022-06-23T10:00:00+02:00",
  "end": "2022-08-23T00:00:00+02:00",
  "eventDurationEditable": true,
  "eventStartEditable": true,
},
```

how can recurring events integrate this way with the "normal" events?

Reply:
I think we can't integrate recurring events with the normal events because in recurring events we have to pass a range of dates.

These recurring events are quite challenging so, we can focus on only one endpoint at a time.

Reply:
having to endpoints is totally fine. but can you adjust the keys and values that are currently there and add the keys and values that are missing from the corresponding calendar base entry that was created?

Reply:
Okay @donni106
We will adjust the keys and values in response

Reply:
is this something you can do soon?

Reply:
Hi @donni106 
We will look into this.

Reply:
Hi @donni106 
We updated the keys and values as you mentioned.

Note: When we update an entry of a recurring event then that updated entry is saved as new entry. but it has same uid as parent event. So, for this we are adding a key (parent_id) which will be show the id of the original event.

And in 'id' key we are showing the id of updated event entry.

Reply:
can you show a json example for that case please

Reply:
For Example, we updated an entry of recurring event like color, title, description 

**See the screenshot-1**
![image](https://user-images.githubusercontent.com/103625795/188837666-9a7d73d7-dc20-41a4-9370-e6462cda4a8d.png)
 
**See the screenshot-2**
![image](https://user-images.githubusercontent.com/103625795/188837159-52464485-246b-4fc5-91b6-b71bfa2e33b2.png)

**And the new entry(id=41) will be saved with the same uid**

![image](https://user-images.githubusercontent.com/103625795/188838132-398c892b-2b76-4110-af5b-7b5d522d90fd.png)

**JSON DATA**

(This is an updated event whose parent_id = 24)

![image](https://user-images.githubusercontent.com/103625795/188838697-05e736ee-717f-41f8-b030-940e63be8193.png)

(And the Id of this updated event is = 41)

![image](https://user-images.githubusercontent.com/103625795/188839015-3c98dd28-9eaa-4869-ac05-be36e08264ca.png)

Reply:
ok thanks so far.
there comes a few more questions to my mind. what data is the one with the id 24? and is the recurring event with id 40 still returned by the api next to the 41?

Reply:
Hi @donni106 
The data which has id = 24 is the main event that is recurring.

Let's take an example: Suppose you created an event( id = 24) and this event is recurring 10 times. But after some time you changed two entries' data of the event (id = 24) like their title, description, etc. So, in 'calendar_entry' table two new entries will be created. 

But the uid of these two entries will be the same as the uid of the main event(id= 24) because these two entries is the part/associated with the event (id=24).

**See the parent id column**

![image](https://user-images.githubusercontent.com/103625795/188853883-f9aeb384-a4d3-491b-9906-cdb9a58644e4.png)

So, when we will fetching this event (id = 24), the 'id' key will be '40' and '41' for 2 entries, and for the remaining 8 entries, It will be null because they are not modified/updated.

Reply:
but only the last entry is correct, am i right? in our case the 41. the 40 is outdated and should not be used?

Reply:
No, check the start_datetime

![image](https://user-images.githubusercontent.com/103625795/188860970-85041ee5-87e1-4fdd-a887-87283d223a79.png)

Reply:
so both entries are valid to show in a list?

Reply:
all right, i think i got it now :)